### ISSUE TITLE
Repackage vcpkg packages as build2 packages

### ISSUE THREAD

**Issue by boris-kolpackov** [build2/build2#35]
Explore the possibility of repackaging `vcpkg` packages as `build2` packages.

First some background: `vcpkg` uses CMake as its underlying build system. For packages that use other build system (for example, Boost), it wraps them with a CMake-based interface. So, in essence, `vcpkg` has a (presumably) uniform, CMake-based interface for building all its packages.

One way such a repackaging could work is by taking these CMake-based packages and making `build2` invoke CMake to build them. More specifically, we could implement an alternative build system rule that could link a library using CMake (*as-if* out of thin air) in addition to an existing link rule that could link a library our of object files. Such a rule will need to translate the `build2` configuration (compiler, options, etc) to CMake's and then "translate" the resulting library/headers back to what other `build2` projects could consume. What exactly these translations will involve remains to be understood. For more details on this idea see the [earlier discussion](https://lists.build2.org/archives/users/2018-November/000494.html).

To answer the obvious question of why not to use `vcpkg` but go straight to CMake: We want to reduce the number of intermediaries that have a chance to reduce the reliability of the combined system. In other words, we want to go to the source, which should also give us maximum control. Regarding `vcpkg` specifically, last time I tried to build it, it did not compile on half of the platforms we are interested in supporting (most notably Mac OS). The same goes for shared libraries on non-Windows platforms. Also, `vcpkg` works in terms of *pre-resolved package version sets* where you move from one set to another rather then upgrading/downgrading individual packages. This does not fit the `build2` model well.

**Comment by GunpowderGuy** [comment]
@boris-kolpackov Has there been any developement on this?

**Comment by boris-kolpackov** [comment]
I believe @Klaim is working on a prototype of a tool that would allow exporting information necessary to use a library from CMake.

**Comment by Klaim** [comment]
Yes, although I couldn't progress on it recently (because dayjob workload vs time).
It's there if you want to take a look: https://github.com/klaim/wyvern
See [this board](https://github.com/Klaim/wyvern/projects/1) for what's missing. I think the notes in klaim-notes.md are not up to date.

**Comment by GunpowderGuy** [comment]
meson forgoed cmake compatibility, what downsided will there be using from encouraging build 2 users to utilize libraries that depend on cmake

**Comment by Klaim** [comment]
> meson forwent cmake compatibility, what downsides will there from encouraging build 2 users to utilize libraries that depend on cmake?

I don't understand your question?

**Comment by GunpowderGuy** [comment]
There will be downsides from encouraging build2 packages to depend on libraries that use CMake. How severe will they be?

**Comment by Klaim** [comment]
Well assuming (lets call that feature A) the prototype I worked on leads to build2 being able to use CMake installed projects (to be clear: the CMake proejct would have to ahve been built before-hand and installed somewhere, then that location passed to build2), feature which would relies on using pkg-config definitions generated from reading these CMake installed projects package files, you would get the same limitations as today when using any pkg-config project from a build2 project:
 - no automatic dependency handling, checks etc. of that package's and it's dependencies;
 - your different configurations would probabily need to use different installs of that package;
 - if you don't have the sources of the CMake project, you would not be able to fix issues in that dependency;
 
So basically similar to having an system-installed package, with maybe just an extra step to specify where to locate the CMake installed package.
 
Then if we are going further (lets call that feature B), by having some ways to automatically call vcpkg to acquire and build these libraries, then from the recent announcements of vcpkg I'm guessing it would be possible to remove all these issues, by letting vcpkg handling dependencies of CMake packages and just providing the necessary info to build2. (Here one potential trouble is that it would remove the need to package anything in build2? depends on the differences with vcpkg...)
 
The main issue here is actually lack of time/resources (as with most open-source projects not pushed by big companies): the build2 team is already hands full with build2 itself, I try to help but my free time is limited and not really predictable, so any help is welcome to push further (a few people already helped a lot with some packages, see https://github.com/build2/build2-packaging/).

Reaching feature A seems very close based on the prototype I did. It lacks the "generate pkg-config files" part but that doesn't seem like a big problem, just need some time to dev and make sure it works with complex cases. That project then would need to be made more maintainable/cleaned up to be useful in production and/or integrated inside build2. Unfortunately I don't have time right now to focus on it, maybe in the summer.

Once feature A is possible, reaching feature B seems possible to do but requires quite some time to get right, in particular because we would need to make sure it's easy to update relative to changes in vcpkg. Though maybe it's not possible to do it right and it would be better to let the user get packages through vcpkg manually and we just provide a simple way to use these packages. Handling multiple configurations that way will probably be a bit messy/cumbersome/error-prone, but that would work enough for me for example.

**Comment by eli-schwartz** [comment]
Probably the biggest issue with cmake interop is that if you want to use the project's export interface, a *-config.cmake file is written in the cmake programming language and cannot be parsed by anything other than genuine cmake.

> meson forwent cmake compatibility, what downsides will there from encouraging build 2 users to utilize libraries that depend on cmake?

Meson kind of has cmake compatibility, in that if you have cmake installed on the system it can execute cmake, then interpret the debug logs from cmake in order to infer:
- build rules, for cmake projects
- link interfaces, for cmake system-installed dependencies via *-config.cmake

This is cautioned to be on a best-effort basis, and of course it does require one to install both cmake and Meson. I suppose build2 could do this, and then require installing both cmake and build2.

What you get out of this is the ability (if everything works correctly) to do a unified build, thus scoring on every point raised above.

> you would get the same limitations as today when using any pkg-config project from a build2 project:
> 
>     * no automatic dependency handling, checks etc. of that package's and it's dependencies;

But, BTW: pkg-config *does* assert that dependencies (with version support) are also available. Also the --static argument to pkg-config will yield the static libraries from all those dependencies, recursively.

**Comment by Klaim** [comment]
> Probably the biggest issue with cmake interop is that if you want to use the project's export interface, a *-config.cmake file is written in the cmake programming language and cannot be parsed by anything other than genuine cmake.

That's why in my attempt to gather these data (https://github.com/klaim/wyvern) I went with running cmake and using the file-api to get the information, instead of trying to read cmake files or logs. It's good enough to consume cmake-generated packages, it's just not equivalent to the information in the cmake scripts (which contains targets that might not appear in the file-api for example).

**Comment by eli-schwartz** [comment]
> That's why in my attempt to gather these data (https://github.com/klaim/wyvern) I went with running cmake and using the file-api to get the information, instead of trying to read cmake files or logs.

Ah, yeah -- that's the same nitty-gritty details Meson uses (the file API is basically debug logs, you can't change my mind :D).

It's painful, good enough to consume packages, and rather flaky when trying to get the information from the cmake scripts.

If you're interested in expanding it past a proof of concept, you might want to take inspiration from the python implementation here: https://github.com/mesonbuild/meson/tree/master/mesonbuild/cmake

**Comment by Klaim** [comment]
> It's painful, good enough to consume packages, and rather flaky when trying to get the information from the cmake scripts.

That summarizes my observations. :)

> If you're interested in expanding it past a proof of concept, you might want to take inspiration from the python implementation here: https://github.com/mesonbuild/meson/tree/master/mesonbuild/cmake

We did look at meson's code before I went with the file-api. Boris pointed at the time that just being able to consume a built cmake project would already be good. Being able to invoke cmake or vcpkg to build that project is kind of a bonus.

As for that project, we had a discussion last year which lead to moving it in stand-by, mainly because it's not really a priority anymore. It's more useful to work on proper `build2` packages. Now `Boost` is packaged and soon `Qt` will be, some other important ones are on the way (SDL and SFML are important to me for example) and even if it's harder to setup, we'll benefit from having proper targets, options, and automatic downloads.

**Comment by boris-kolpackov** [comment]
While I no longer believe repackaging `vcpkg` as `build2` packages is a good idea (because they will always be second-class citizens), it might still be a good idea to consider integration with `vcpkg` as a system package manager, similar to the integration we are currently working on for the `deb` and `rpm` package managers.

**Comment by helmesjo** [comment]
Yeah wrapping `CMake` in `build2` sounds like a haunting rabbit hole, so let's avoid the `vcpkg`/`conan` approach (it's a mess for other teams around me already).
